### ISSUE TITLE
Investigation for Vaisala

### ISSUE THREAD

**Issue by simboss** [geosolutions-it/aaime#43]
From the client: 

It’s been awhile since we spoke during “raster crossing the dateline” work you did for us. Quite a lot has happened since, but GIS is still one of the important domains for us. We want to offer a consistent look and feel for maps in both online and offline products. To achieve that, we have chosen maptiler.org for hosting online maps, OpenMapTiles as the offline dataset and MapBox GL JS as the style definition language.

 

There is problem we haven’t found a solution for yet; we need to offer maps in projections other than Spherical Mercator. Specifically, some of our offline products need to support Azimuthal Equidistant projections at various locations all over the Earth. The existing software accesses raster tiles from GeoServer over WMS. We would prefer not to change the interface from WMS, and the projection definitions used are currently specific to GeoServer, so a GeoServer-based solution would be preferred.

 

In short, we are looking for a PoC of GeoServer rendering OpenMapTiles data with a custom MapBox Gl JS style (our style available on request) in arbitrary projections. Is this something you could help us and if so how would you like to proceed?

We need to do an investigation and provide solution + estimates.

**Comment by aaime** [comment]
So do we need to generate vector tiles in another projection and then have rendering happen client side?
Or do we need to render server side? The MBStyles from OpenMapTiles are using a recent version of the spec and are not usable in GeoServer unfortunately (hence, if we need to render server side, an estimate of upgrading style compact or developing our own in CSS/SLD would be needed).

**Comment by simboss** [comment]
I would go ahead and estimate everything.

Then I will decide if we take something out or if I show them everything.

**Comment by aaime** [comment]
Checked how the OpenMapTiles machinery works. The system is basically split in two, there is a set of tools that are designed to build a mbtiles file with the vector tiles, and another to serve it.
The tools that build the vector tiles:

- import data from various sources (OSM shapefiles, natural earth, OSM pbf files) into a postgis database using imposm
- build a set of tables and views (eventually materialized) per zoom level, and then sql functions on top of them taking a bbox and a zoom level as a parameter, the latter are actually used to fetch data from
- create the mbtiles by calling the functions with the right bbox and zoom level for each tile

Once the mbtiles is done, the above gets shut down (it's a set of docker containers) and another container is started with a node.js process that:
- servers mbtiles on request
- can build PNG tiles too by rendering the vector tiles using javascript and I guess Mapbox GL code

Long story short, first one builds a fixed set of tiles that are independent of the styles, and only driven by the filters contained in the SQL definitions, and then it's possible to apply styling on top of it.
Given that the database is geared towards zoom level based usage, and for one-off generation of tiles, it's not immediate to use it for general serving. In particular, the materialized views would have to be rebuilt/refreshed in case of data update (they do not update automatically)... the OpenMapTiles import process actually drops and re-creates them, so it would have to be tweaked, or it would break a GeoServer running against the database.

Styling wise, GeoTools is not able to read the styles used by OpenMapTiles, as they use newer functionality. I don't know how much the two deviate, unfortunately the translation dies at the first problem found.

Here are some ROM estimates:

- Building a data dir that can use the OpenMapTiles database as a source, and tweaks to import procedure so that it does not wipe out stuff we depend onto: 5 days
- Manually clone one of the OpenMapTiles styles using GeoCSS: 10 days per style? (checked with Stefano, we have the same feeling, of course, it's just a very rough guess)
- Fix the mbtiles reader to read the styles... I have no idea :-D . If I could spend a day on it to read the specs and figure out what the geotools modules cannot do that the spec has, I could give you a better estimate. For now, if I had to throw a number, I'd think 10 days would allow to do a lot code wise.

Another possibly "funky" idea would be to just use the mbtiles generated by OpenMapTiles as a vector data source, building a datastore around it, and re-assembling polygon/lines on the fly, picking the best source based on the generalization hint that the renderer provides based on the output resolution.
Re-assembling polygons and lines can be pretty tricky though, and would have to go through JTS.
We'd also need some way to figure out which layers are contained in the tiles, I'm guessing some inspection/configuration would be needed (maybe a quick scan generating a template config file that someone can tweak by hand). There would also be some efficiency concern, as the renderer works by layer, meaning the same tile would have to be read multiple times to fetch the desired layers out of it (maybe some caching could be used).
I would guess a store like this could be coded in around 10 days.

**Comment by simboss** [comment]
@aaime about the funky idea, would this play nice with showing up polar areas? Moreover, they explicitly mention Azimuthal Equidistant projections, again would they play nice with this approach.

What about the other, longer, approach?

**Comment by aaime** [comment]
The databases for OpenMapTiles are all in 3857, so they cannot contains any data over the polar areas. Hence, there should not be reprojection issues.

**Comment by simboss** [comment]
K, we should make this clean with the client then.

**Comment by simboss** [comment]
I kind of like option 2, how confident are you with this solution?
Do we need to raise the estimate a little?

**Comment by simboss** [comment]
Feedback from client

==
Concentrating on the technical side here.

Yes, we need server-side rendering. The idea is to have this as drop-in replacement for an existing WMS base map service to support legacy clients that can't be changed significantly.

We would thus need to ensure that the our MBStyles rendered in GeoServer produce results similar to rendering OMT data with openmaptiles-server. We don’t expect to use GeoServer-rendered raster tiles in the same context as vector tiles, so we don’t expect exactly the same rendering results, but we want to avoid redefining the styles in a different style language, or for a different data schema.

As for performance, we expect to cache the rendered raster tiles, so tile rendering performance is not the primary concern. Of course we still need to populate the cache, but we wouldn't expect raster tiles to be rendered on-the-fly as the user zooms and pans around.

Data-wise, we would like to have a process that starts with the OMT MBTiles files. We would also like to avoid shipping PostGIS databases around, so if an intermediate format is required, we would rather use GeoPackage or something else that can be shipped around the world without a lengthy restore process at the deployment site. We understand that the MBTiles are latitude-clipped because of the Spherical Mercator projection, but currently don’t see this as an issue.

When it comes to approach -1-, it seems to me that simply reading the (tile-clipped) vector features from MBTiles and rendering them in GeoServer using an MBStyle should be sufficient to replicate the rendering results of MBGLJS. That is, we don’t particularly want to reconstruct the original features from the clipped features, unless it is actually necessary for rendering it to raster image. If simply reading the MBTiles in is not sufficient for MBStyles-based rendering in GeoServer, it would perhaps be better to try to acquire the data in some other schema-compatible format to begin with, or to produce a separate tool for the data conversion.
==

What do you think we need in order to refine the estimates?
They seem to be willing to move forward so I would want to close the deal sooner rather than later.

**Comment by aaime** [comment]
Quick thoughs before I can do a deeper investigation:
* We have to merge the geometries back toghether prior to rendering, or we'll end up with rendering artifacts, very visible on polygons, less so with lines but still visible. However I could add a rendering hint that the renderer only wants the part of the data inside the rendering area (a sort of a rendering clip bbox, which is normally larger than the rendered area) 
* We'll have to check which parts of the OpenMapStyles are not supported by the current geotools module and do an estimate (right now it breaks at the first issue found, need to change the code so that it keeps on going on and we can figure out how many fixes are needed)

**Comment by simboss** [comment]
I don't understand what you need to move forward.

**Comment by aaime** [comment]
I've looked into parsing the OpenMapStyle styles, in particular, the dark-matter-gl-style,  fiord-color-gl-style,  klokantech-basic-gl-style,  osm-bright-gl-style,  positron-gl-style and  toner-gl-style and figured out what it would take to parse them without crashes.

Then looked at the outputs of a couple in terms of SLD, manually comparing with the MBStyle source, and did some estimates to make the output actually comply with the MBStyle, added some estimates here too, as well as some padding for the unforeseen part.

Finally estimated the mbtiles + vectortiles datastore work.

Everything is available here: https://docs.google.com/spreadsheets/d/13HA-UI2WuP1jfzbHdZzP-4W-tqq8y4WvVH5GfZkU9VY/edit#gid=376642658

**Comment by simboss** [comment]
work accepted
